File: java/Disruptor/src/main/java/cn/wjybxx/disruptor/RingBufferEventSequencer.java
/*
 * Copyright 2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.disruptor;

import java.util.Objects;
import java.util.concurrent.TimeUnit;

/**
 * @author wjybxx
 * date - 2024/1/17
 */
public class RingBufferEventSequencer<T> implements EventSequencer<T> {

    private final RingBuffer<T> buffer;
    private final RingBufferSequencer sequencer;

    private RingBufferEventSequencer(Builder<T> builder) {
        Objects.requireNonNull(builder);
        this.buffer = new RingBuffer<>(builder.getFactory(), builder.getBufferSize());
        if (builder.getProducerType() == ProducerType.MULTI) {
            sequencer = new MultiProducerSequencer(
                    builder.getBufferSize(),
                    builder.getProducerSleepNanos(),
                    builder.getWaitStrategy(),
                    builder.getBlocker());
        } else {
            sequencer = new SingleProducerSequencer(
                    builder.getBufferSize(),
                    builder.getProducerSleepNanos(),
                    builder.getWaitStrategy(),
                    builder.getBlocker());
        }
    }

    public RingBuffer<T> getBuffer() {
        return buffer;
    }

    // region buffer

    @Override
    public final T get(long sequence) {
        return buffer.getElement(sequence);
    }

    @Override
    public final T producerGet(long sequence) {
        return buffer.getElement(sequence);
    }

    @Override
    public final T consumerGet(long sequence) {
        return buffer.getElement(sequence);
    }

    @Override
    public void producerSet(long sequence, T data) {
        Objects.requireNonNull(data);
        buffer.setElement(sequence, data);
    }

    @Override
    public void consumerSet(long sequence, T data) {
        buffer.setElement(sequence, data);
    }

    @Override
    public int capacity() {
        return buffer.getBufferSize();
    }

    @Override
    public long remainingCapacity() {
        return sequencer.remainingCapacity();
    }

    @Override
    public Sequencer sequencer() {
        return sequencer;
    }

    @Override
    public ProducerBarrier producerBarrier() {
        return sequencer;
    }

    @Override
    public DataProvider<T> dataProvider() {
        return buffer;
    }
    // endregion

    // region producer

    @Override
    public boolean hasAvailableCapacity(int requiredCapacity) {
        return sequencer.hasAvailableCapacity(requiredCapacity);
    }

    @Override
    public long next() {
        return sequencer.next(1); // 传入1可减少调用
    }

    @Override
    public long next(int n) {
        return sequencer.next(n);
    }

    @Override
    public long tryNext() {
        return sequencer.tryNext(1);
    }

    @Override
    public long tryNext(int n) {
        return sequencer.tryNext(n);
    }

    @Override
    public long nextInterruptibly() throws InterruptedException {
        return sequencer.nextInterruptibly(1);
    }

    @Override
    public long nextInterruptibly(int n) throws InterruptedException {
        return sequencer.nextInterruptibly(n);
    }

    @Override
    public long tryNext(int n, long timeout, TimeUnit unit) {
        return sequencer.tryNext(n, timeout, unit);
    }

    @Override
    public void publish(long sequence) {
        sequencer.publish(sequence);
    }

    @Override
    public void publish(long lo, long hi) {
        sequencer.publish(lo, hi);
    }
    // endregion

    // region builder

    /** 多线程生产者builder */
    public static <T> Builder<T> newMultiProducer(EventFactory<? extends T> factory) {
        return new Builder<T>(factory)
                .setProducerType(ProducerType.MULTI);
    }

    /** 单线程生产者builder */
    public static <T> Builder<T> newSingleProducer(EventFactory<? extends T> factory) {
        return new Builder<T>(factory)
                .setProducerType(ProducerType.SINGLE);
    }

    public static class Builder<T> extends EventSequencerBuilder<T> {

        private ProducerType producerType = ProducerType.MULTI;
        private int bufferSize = 8192;

        public Builder(EventFactory<? extends T> factory) {
            super(factory);
        }

        @Override
        public RingBufferEventSequencer<T> build() {
            return new RingBufferEventSequencer<>(this);
        }

        /** 生产者的类型 */
        public ProducerType getProducerType() {
            return producerType;
        }

        public Builder<T> setProducerType(ProducerType producerType) {
            this.producerType = Objects.requireNonNull(producerType);
            return this;
        }

        /** 环形缓冲区的大小 */
        public int getBufferSize() {
            return bufferSize;
        }

        public Builder<T> setBufferSize(int bufferSize) {
            this.bufferSize = bufferSize;
            return this;
        }

        @Override
        public Builder<T> setProducerSleepNanos(long producerSleepNanos) {
            return (Builder<T>) super.setProducerSleepNanos(producerSleepNanos);
        }

        @Override
        public Builder<T> setWaitStrategy(WaitStrategy waitStrategy) {
            return (Builder<T>) super.setWaitStrategy(waitStrategy);
        }

        @Override
        public Builder<T> enableBlocker() {
            return (Builder<T>) super.enableBlocker();
        }

        @Override
        public Builder<T> disableBlocker() {
            return (Builder<T>) super.disableBlocker();
        }
    }

    // endregion
}

File: csharp/Wjybxx.Disruptor/src/RingBufferEventSequencer.cs
#region LICENSE

// Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;

namespace Wjybxx.Disruptor
{
public class RingBufferEventSequencer<T> : EventSequencer<T>
{
    private readonly RingBuffer<T> buffer;
    private readonly RingBufferSequencer _sequencer;

    private RingBufferEventSequencer(Builder builder) {
        if (builder == null) throw new ArgumentNullException(nameof(builder));
        this.buffer = new RingBuffer<T>(builder.Factory, builder.BufferLength);
        if (builder.ProducerType == ProducerType.Multi) {
            _sequencer = new MultiProducerSequencer(
                builder.BufferLength,
                builder.ProducerSpinIterations,
                builder.WaitStrategy,
                builder.Blocker);
        } else {
            _sequencer = new SingleProducerSequencer(
                builder.BufferLength,
                builder.ProducerSpinIterations,
                builder.WaitStrategy,
                builder.Blocker);
        }
    }

    public RingBuffer<T> Buffer => buffer;

    #region buffer

    public T Get(long sequence) {
        return buffer.GetElement(sequence);
    }

    public T ProducerGet(long sequence) {
        return buffer.GetElement(sequence);
    }

    public T ConsumerGet(long sequence) {
        return buffer.GetElement(sequence);
    }

    public void ProducerSet(long sequence, T data) {
        buffer.SetElement(sequence, data);
    }

    public void ConsumerSet(long sequence, T data) {
        buffer.SetElement(sequence, data);
    }

    public ref T ProducerGetRef(long sequence) {
        return ref buffer.GetElementRef(sequence);
    }

    public ref T ConsumerGetRef(long sequence) {
        return ref buffer.GetElementRef(sequence);
    }

    public int Capacity => buffer.BufferLength;

    public long RemainingCapacity => _sequencer.RemainingCapacity();

    public Sequencer Sequencer => _sequencer;

    public ProducerBarrier ProducerBarrier => _sequencer;

    public DataProvider<T> DataProvider => buffer;

    #endregion

    #region producer

    public bool HasAvailableCapacity(int requiredCapacity) {
        return _sequencer.HasAvailableCapacity(requiredCapacity);
    }

    public long Next() {
        return _sequencer.Next(1); // 传入1可减少调用
    }

    public long Next(int n) {
        return _sequencer.Next(n);
    }

    public long? TryNext() {
        return _sequencer.TryNext(1);
    }

    public long? TryNext(int n) {
        return _sequencer.TryNext(n);
    }

    public long NextInterruptibly() {
        return _sequencer.NextInterruptibly(1);
    }

    public long NextInterruptibly(int n) {
        return _sequencer.NextInterruptibly(n);
    }

    public long? TryNext(int n, TimeSpan timeout) {
        return _sequencer.TryNext(n, timeout);
    }

    public void Publish(long sequence) {
        _sequencer.Publish(sequence);
    }

    public void Publish(long lo, long hi) {
        _sequencer.Publish(lo, hi);
    }

    #endregion

    #region builder

    /** 多线程生产者builder */
    public static Builder NewMultiProducer(Func<T> factory) {
        return new Builder(factory)
        {
            ProducerType = ProducerType.Multi
        };
    }

    /** 单线程生产者builder */
    public static Builder NewSingleProducer(Func<T> factory) {
        return new Builder(factory)
        {
            ProducerType = ProducerType.Single
        };
    }

    public class Builder : EventSequencerBuilder<T>
    {
        private ProducerType producerType = ProducerType.Multi;
        private int bufferLength = 8192;

        public Builder(Func<T> factory) : base(factory) {
        }

#if NET5_0_OR_GREATER
        public override RingBufferEventSequencer<T> Build() {
            return new RingBufferEventSequencer<T>(this);
        }
#else
        public override EventSequencer<T> Build() {
            return new RingBufferEventSequencer<T>(this);
        }
#endif

        /// <summary>
        /// 生产者的类型
        /// </summary>
        public ProducerType ProducerType {
            get => producerType;
            set => producerType = value;
        }

        /// <summary>
        /// 环形缓冲区的大小
        /// </summary>
        public int BufferLength {
            get => bufferLength;
            set => bufferLength = value;
        }
    }

    #endregion
}
}